core/lib/dal/src/contract_verification_dal.rs

```diff
@@ -560,6 +583,285 @@ impl ContractVerificationDal<'_, '_> {
+                bytecode_without_metadata_sha3.as_bytes()
+            )
+            .try_map(|row| {
+                let info = serde_json::from_value::<VerificationInfo>(row.verification_info)
+                    .decode_column("verification_info")?;
+                let bytecode_keccak256 = H256::from_slice(&row.bytecode_keccak256);
+                let bytecode_without_metadata_keccak256 = row
+                    .bytecode_without_metadata_keccak256
+                    .map(|h| H256::from_slice(&h));
+                Ok((
+                    info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256,
+                ))
+            })
+            .instrument("get_contract_verification_info_v2#perfect_match")
+            .with_arg("address", &address)
+            .fetch_optional(self.storage)
+            .await?
+        else {
+            return Ok(None);
+        };
+        if identifier.bytecode_sha3 != bytecode_keccak256 {
+            // Sanity check
+            if bytecode_without_metadata_keccak256.is_none()
+                || identifier.bytecode_without_metadata_sha3.map(|h| h.hash())
+                    != bytecode_without_metadata_keccak256
+            {
+                tracing::error!(
+                    contract_address = %address,
+                    identifier = ?identifier,
+                    bytecode_keccak256 = %bytecode_keccak256,
+                    info = ?info,
+                    "Bogus verification info fetched for contract",
+                );
+                anyhow::bail!("Internal error: bogus verification info detected");
+            }
+
+            // Mark the contract as partial match (regardless of other issues).
+            info.verification_problems = vec![VerificationProblem::IncorrectMetadata];
+        }
+        Ok(Some(info))
+    }
+
+    /// Checks if migration from `contracts_verification_info` to `contract_verification_info_v2` is performed
+    /// by checking if the latter has more or equal number of rows.
+    pub async fn is_verification_info_migration_performed(&mut self) -> DalResult<bool> {
+        let count_v1 = sqlx::query!(
+            r#"
+            SELECT
+                COUNT(*)
+            FROM
+                contracts_verification_info
+            "#,
+        )
+        .instrument("is_verification_info_migration_performed#count_v1")
+        .fetch_one(self.storage)
+        .await?
+        .count
+        .unwrap() as usize;
+
+        let count_v2 = sqlx::query!(
+            r#"
+            SELECT
+                COUNT(*)
+            FROM
+                contract_verification_info_v2
+            "#,
+        )
+        .instrument("is_verification_info_migration_performed#count_v2")
+        .fetch_one(self.storage)
+        .await?
+        .count
+        .unwrap() as usize;
+
+        Ok(count_v2 >= count_v1)
+    }
+
+    pub async fn perform_verification_info_migration(
```

Comment:
Bikeshedding: Subjectively, this mixes DB queries with high-level logic. I'd prefer the latter to be moved to the contract verifier library.

Reply:
The logic here was that it's a one-time thing that should be removed after the migration is done everywhere, so it would be easier to remove it if it exists in a single place. Otherwise, we can get some dead leftovers. WDYT?

Reply:
IMO, it would be manageable to isolate the temporary DB queries into a separate `impl` block with a `deprecated` attribute. This worked for some previous migrations.

Reply:
I've moved some of the logic out of migration, but otherwise I think it's OK to keep as-is. Now it's _mostly_ SQL, and moreover everything is performed in a single long-running transaction.